### PR TITLE
CBG-2842 Always return unauthorized for non authorized public API

### PR DESCRIPTION
**Pull request by torcolvin**
## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration-1.19.5/4/

### CODE REVIEW

File: rest/api_test.go [hunk 1]
```diff
@@ -217,29 +217,33 @@ func TestCORSOrigin(t *testing.T) {
 	reqHeaders := map[string]string{
 		"Origin": "http://example.com",
 	}
-	response := rt.SendRequestWithHeaders("GET", "/{{.keyspace}}/", "", reqHeaders)
+	response := rt.SendRequestWithHeaders("GET", "/{{.db}}/", "", reqHeaders)
+	RequireStatus(t, response, http.StatusUnauthorized)
```
Comment (adamcfraser) [inline rest/api_test.go]: Were these tests previously returning these statuses, or does that change in this PR?

Comment (adamcfraser) [inline rest/api_test.go]: Were these tests previously returning these statuses, or does that change in this PR?

File: rest/user_api_test.go [hunk 2]
```diff
@@ -1520,3 +1520,56 @@ func TestPutUserCollectionAccess(t *testing.T) {
...
+		PersistentConfig: true,
+	})
+	defer rt.Close()
+
+	dbConfig := rt.NewDbConfig()
+
+	dbName := "realdb"
+	rt.CreateDatabase("realdb", dbConfig)
+
+	response := rt.SendRequest(http.MethodGet, "/{{.db}}/", "")
```
Comment (adamcfraser) [inline rest/user_api_test.go]: Does this resolve to "realdb"?  Is there an advantage to using the token here - to me it seems to obscure what's actually being tested.

File: rest/handler.go [hunk 3]
```diff
@@ -688,6 +695,11 @@ func checkJWTIssuerStillValid(ctx context.Context, dbCtx *db.DatabaseContext, us
 // checkAdminAuthenticationOnly simply checks whether a username / password combination is authenticated pulling the
 // credentials from the handler
 func (h *handler) checkAdminAuthenticationOnly() (bool, error) {
+	if h.privs == adminPrivs && !*h.server.Config.API.AdminInterfaceAuthentication {
```
Comment (adamcfraser) [inline rest/handler.go]: Is this change required?  This looks like it's already being evaluated by the called (if shouldCheckAdminAuth).

**Review by adamcfraser — COMMENTED** [review]
The PR description describes this as handling the admin case, but the associated ticket (CBG-2842) is for public access.  Which case is this fixing?

File: rest/api_test.go (hunk 1 above)
Comment (torcolvin) [reply rest/api_test.go]: If you ask `{{.keyspace}}` it will return 400, which is how keyspace is resolved than db differently. I could leave this and test that it returns 400. 

This change is only a test clarification, I wanted to assert I wasn't changing the error code, but I can leave this out of this initial PR.

I added this code to make sure nothing was changed.

Comment (torcolvin) [reply rest/api_test.go]: Yes, using the same.

File: rest/user_api_test.go (hunk 2 above)
Comment (torcolvin) [reply rest/user_api_test.go]: yeah, fixing.

**Comment by torcolvin** [comment]
> The PR description describes this as handling the admin case, but the associated ticket ([CBG-2842](https://issues.couchbase.com/browse/CBG-2842)) is for public access. Which case is this fixing?

This was just a mistake of a message, this fixes public access and doesn't regress on admin access.

File: rest/handler.go (hunk 3 above)
Comment (torcolvin) [reply rest/handler.go]: nope

File: rest/api_test.go [hunk 4]
```diff
@@ -114,11 +120,22 @@ func TestDisablePublicBasicAuth(t *testing.T) {
...
+
+	response = rt.SendUserRequest(http.MethodGet, "/notadb/", "", "user1")
+	RequireStatus(t, response, http.StatusUnauthorized)
+	require.Contains(t, response.Body.String(), ErrInvalidLogin.Message)
 	assert.NotContains(t, response.Header(), "WWW-Authenticate", "expected to not receive a WWW-Auth header when password auth is disabled")
 
 	// Also check that we can't create a session through POST /db/_session
 	response = rt.SendRequest(http.MethodPost, "/{{.db}}/_session", `{"name":"user1","password":"letmein"}`)
 	RequireStatus(t, response, http.StatusUnauthorized)
+	require.Contains(t, response.Body.String(), ErrLoginRequired.Message)
```
Comment (adamcfraser) [inline rest/api_test.go:134]: I'm not sure whether this is testing what it wants to be testing - it looks like the "Login required" might be happening during authentication of the request (if it's trying to auth against the public port as GUEST).  I think the intention here is to test that DisablePublicBasicAuth prevents creating a session using username/password in the session body.  So we want the request to authenticate successfully, but the session creation to fail.

This looks like an existing test bug - maybe file a non-Helium ticket to correct this test?

File: rest/handler.go [hunk 5]
```diff
@@ -1049,6 +1067,21 @@ func (h *handler) getBearerToken() string {
 	return ""
 }
 
+// providedAuthCredentials returns true if basic auth or session auth is enabled
+func (h *handler) providedAuthCredentionals() bool {
```
Comment (adamcfraser) [inline rest/handler.go]:
```suggestion
func (h *handler) providedAuthCredentials() bool {
```

File: rest/handler.go [hunk 6]
```diff
@@ -1049,6 +1067,21 @@ func (h *handler) getBearerToken() string {
 	return ""
 }
 
+// providedAuthCredentials returns true if basic auth or session auth is enabled
```
Comment (adamcfraser) [inline rest/handler.go:1070]: This doesn't cover all the mechanisms by which auth could be provided (it's missing bearer JWT token).  The session cookie check is also only going to work in cases where a valid db has been found, which doesn't seem like it's necessarily the case for all callers.

File: rest/handler.go [hunk 7]
```diff
@@ -648,7 +664,10 @@ func (h *handler) checkAuth(dbCtx *db.DatabaseContext) (err error) {
 		if dbCtx.Options.SendWWWAuthenticateHeader == nil || *dbCtx.Options.SendWWWAuthenticateHeader {
 			h.response.Header().Set("WWW-Authenticate", wwwAuthenticateHeader)
 		}
-		return base.HTTPErrorf(http.StatusUnauthorized, "Login required")
+		if h.providedAuthCredentionals() {
```
Comment (adamcfraser) [inline rest/handler.go]: It looks to me like we've already fallen through to GUEST access handling at this point, which I thought would imply that auth credentials haven't been provided.  Which scenario is this covering?

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/handler.go (hunk 6 above)
Comment (torcolvin) [reply rest/handler.go:1070]: It's not the case for all callers, which is why the check only works if it's the default cookie type if database is fine. I added bearer check though.

File: rest/handler.go (hunk 7 above)
Comment (torcolvin) [reply rest/handler.go]: This is hit by 

```
        response = rt.SendUserRequest(http.MethodGet, "/{{.db}}/", "", "user1")
	RequireStatus(t, response, http.StatusUnauthorized)
	require.Contains(t, response.Body.String(), ErrInvalidLogin.Message)
```

from `TestDisablePublicBasicAuth`

File: rest/api_test.go (hunk 4 above)
Comment (torcolvin) [reply rest/api_test.go:134]: This is actually behavior I changed - 

Otherwise:

```response = rt.SendRequest(http.MethodPost, "/{{.db}}/_session", `{"name":"user1","password":"letmein"}`)```

and 

```response = rt.SendRequest(http.MethodPost, "/notadb/_session", `{"name":"user1","password":"letmein"}`)```

would return different errors. The change is in session_api.go

File: rest/handler.go (hunk 7 above)
Comment (adamcfraser) [reply rest/handler.go]: I thought it would be preferable to do this check earlier in checkAuth (i.e. if DisablePasswordAuth is true, return ErrInvalidLogin if basic auth credentials are provided).  However, that would break the existing behaviour, where if guest is enabled AND DisablePasswordAuth is true, basic auth credentials are ignored and it falls through to GUEST.  I'm not convinced this is the desired behaviour, but it's not a security issue and we don't want to break existing handling.

**Review by adamcfraser — APPROVED** [review]
